caffe2/perfkernels/sve_emblookup_codegen.py

```diff
@@ -0,0 +1,469 @@
+        code.append("      if (normalize_by_lengths && length != 0) {")
+        code.append("        const float len_inv = 1.0f / length;")
+    else:
+        exit(1)
+
+    code.append("        svfloat32_t vlen_inv = svdup_n_f32(len_inv);")
+    code.append("        svbool_t pg;")
+    code.append("        for (int64_t j = 0;\n"
+                "             svptest_first(svAll, pg = svwhilelt_b32_s64("
+                "j, block_size));")
+    code.append("             j += vLen) {")
+    code.append(
+        "          svst1_f32(\n"
+        "              pg, &op[j], svmul_f32_x(pg, svld1_f32(pg, &op[j]), vlen_inv));"
+    )
+    code.append("        }")
+    code.append("      }")
+    code.append("    }")
+    return code
+
+
+# start main code
+parser = argparse.ArgumentParser()
+parser.add_argument("-f", "--filename", help="file name")
+parser.add_argument("--fused", action="store_true")
+parser.add_argument("--use-offsets", action="store_true")
+opts = parser.parse_args()
+if opts.filename:
+    filename = opts.filename
+elif opts.fused:
+    print("This script currently does not support --fused")
+    exit(1)
+else:
+    if opts.use_offsets:
+        filename = "embedding_lookup_idx_sve.cc"
+    else:
+        print("This script currently only supports codegen with --use-offsets")
+        exit(1)
+        filename = "embedding_lookup_sve.cc"
+
+options = [
+    ["int32_t", "int32_t", "float", "float", "float", "float"],
+    ["int64_t", "int64_t", "float", "float", "float", "float"],
+    ["int32_t", "int32_t", "half", "at::Half", "float", "float"],
+    ["int64_t", "int64_t", "half", "at::Half", "float", "float"],
+    ["int32_t", "int32_t", "bfloat16", "at::BFloat16", "float", "float"],
+    ["int64_t", "int64_t", "bfloat16", "at::BFloat16", "float", "float"],
+    ["int32_t", "int32_t", "uint8_t", "uint8_t", "float", "float"],
+    ["int64_t", "int64_t", "uint8_t", "uint8_t", "float", "float"],
+]
+
+code = []
+# includes
+code.append("//// --------------------------")
+code.append("//// ATTENTION:")
+code.append("//// THIS CODE IS AUTOGENERATED")
+code.append(f"//// BY {sys.argv[0]}")
+code.append("//// DO NOT MODIFY!!!")
+code.append("//// --------------------------\n")
+
+code.append("#include <arm_sve.h>")
+code.append("#include <c10/util/BFloat16.h>")
+code.append("#include <c10/util/Half.h>")
+code.append("#include <cstdint>")
+code.append("#include <cstring>")
+
+code.append("namespace caffe2 {\n")
+for o in options:
+    [IndexTypeName, IndexType, InTypeName, InType, OutTypeName, OutType] = o
+
+    prefix = "Fused8BitRowwise" if opts.fused else ""
+    code.append("template <bool IS_WEIGHT_POSITIONAL>")
+    if opts.use_offsets:
+        fn_base = f"{prefix}EmbeddingLookupIdx_{IndexTypeName}_{InTypeName}_{OutTypeName}"
+    else:
+        exit(1)
+
+    suffix = "__sve"
+    fn = "static bool " + fn_base + suffix
+    code.append(fn + "(")
+
+    args = []
```

Comment:
why do we need a separate list here?

Reply:
It is reused a few lines down for the weight positional version of the function